openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -23,122 +27,207 @@
         return super(CallStackMixin, self).delete(*args, **kwargs)
 
 
 class CallStackManager(Manager):
-    """ A Manager class which overrides the default Manager class for getting call stacks
+    """ Manager class which overrides the default Manager class for getting call stacks
     """
     def get_query_set(self):
-        """overriding the default queryset API method
+        """ Override the default queryset API method
         """
-        capture_call_stack(self.model)
+        if hasattr(self, 'model'):
+            capture_call_stack(self.model)
+        else:
+            capture_call_stack(type(self))
         return super(CallStackManager, self).get_query_set()
 
 
-def donottrack(*classes_not_to_be_tracked):
-    """function decorator which deals with toggling call stack
-    Args:
-        classes_not_to_be_tracked: model classes where tracking is undesirable
+def donottrack(*entities_not_to_be_tracked):
+    """ Decorator which halts tracking for some entities for specific functions
+
+    Arguments:
+        entities_not_to_be_tracked: entities which are not to be tracked
+
     Returns:
         wrapped function
     """
+    if not entities_not_to_be_tracked:
+        entities_not_to_be_tracked = [None]
 
-    def real_donottrack(function):
-        """takes function to be decorated and returns wrapped function
+    @wrapt.decorator
+    def real_donottrack(wrapped, instance, args, kwargs):  # pylint: disable=W0613
+        """ Takes function to be decorated and returns wrapped function
 
-        Args:
-            function - wrapped function i.e. real_donottrack
+        Arguments:
+            wrapped - The wrapped function which in turns needs to be called by wrapper function.
+            instance - The object to which the wrapped function was bound when it was called.
+            args - The list of positional arguments supplied when the decorated function was called.
+            kwargs - The dictionary of keyword arguments supplied when the decorated function was called.
+
+        Returns:
+            return of wrapped function
         """
-        def wrapper(*args, **kwargs):
-            """ wrapper function for decorated function
-            Returns:
-                wrapper function i.e. wrapper
-            """
-            if len(classes_not_to_be_tracked) == 0:
-                global TRACK_FLAG  # pylint: disable=W0603
-                current_flag = TRACK_FLAG
-                TRACK_FLAG = False
-                function(*args, **kwargs)
-                TRACK_FLAG = current_flag
-            else:
-                global HALT_TRACKING  # pylint: disable=W0603
-                current_halt_track = HALT_TRACKING
-                HALT_TRACKING = classes_not_to_be_tracked
-                function(*args, **kwargs)
-                HALT_TRACKING = current_halt_track
-        return wrapper
+        global HALT_TRACKING  # pylint: disable=W0603
+        HALT_TRACKING.append(entities_not_to_be_tracked)
```

Comment:
You don't need `global` here, since you're just mutating the list, not setting it.